libraries/GCS_MAVLink/GCS_MAVFTP.cpp

```diff
@@ -0,0 +1,713 @@
+            response_payload->data[offset++] = '\0';
+        }
+        else {
+            // Everything else just sends name
+            strncpy((char*)&response_payload->data[offset], result->d_name, remaining_space);
+            response_payload->data[MAX_FTP_DATA_LENGTH - 1] = '\0';
+            offset += MIN(strlen(result->d_name), (unsigned)remaining_space);
+            response_payload->data[offset++] = '\0';
+        }
+    }
+    closedir(dp);
+    response_payload->size = offset;
+    _ftp_errno = errno;
+    return errorCode;
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_Open(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload, int oflag)
+{
+    if (_session_info.fd >= 0) {
+        return kErrNoSessionsAvailable;
+    }
+
+    uint32_t fileSize = 0;
+    struct stat st;
+    _set_current_abs_path(_data_as_cstring(request_payload));
+
+    if (stat(_ftp_current_abs_path, &st)) {
+        // fail only if requested open for read
+        if (oflag & O_RDONLY) {
+            _ftp_errno = errno;
+            return kErrFailErrno;
+
+        }
+        else {
+            st.st_size = 0;
+        }
+    }
+    fileSize = st.st_size;
+
+    int fd = ::open(_ftp_current_abs_path, oflag);
+
+    if (fd < 0) {
+        _ftp_errno = errno;
+        return kErrFailErrno;
+    }
+
+    _session_info.fd = fd;
+    _session_info.file_size = fileSize;
+    _session_info.stream_download = false;
+
+    response_payload->session = 0;
+    response_payload->size = sizeof(uint32_t);
+    memcpy(response_payload->data, &fileSize, response_payload->size);
+    return kErrNone;
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_Read(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    if (request_payload->session != 0 || _session_info.fd < 0) {
+        return kErrInvalidSession;
+    }
+    response_payload->offset = 0;
+    // We have to test seek past EOF ourselves, lseek will allow seek past EOF
+    if (request_payload->offset >= _session_info.file_size) {
+        gcs().send_statustext(MAV_SEVERITY_CRITICAL, (1<<chan), "request past EOF");
+        return kErrEOF;
+    }
+
+    if (lseek(_session_info.fd, request_payload->offset, SEEK_SET) < 0) {
+        gcs().send_statustext(MAV_SEVERITY_CRITICAL, (1<<chan), "seek fail");
+        _ftp_errno = errno;
+        return kErrFailErrno;
+    }
+
+    int bytes_read = ::read(_session_info.fd, &response_payload->data[0], MAX_FTP_DATA_LENGTH);
+
+    if (bytes_read < 0) {
+        // Negative return indicates error other than eof
+        gcs().send_statustext(MAV_SEVERITY_CRITICAL, (1<<chan), "read fail");
+        _ftp_errno = errno;
+        return kErrFailErrno;
+    }
+
+    response_payload->size = bytes_read;
+    response_payload->offset = request_payload->offset;
+    return kErrNone;
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_Burst(FTPPayloadHeader *request_payload, uint8_t target_system_id)
```

Comment:
this function needs to clear some of the response_payload fields
if the lasts response had a size,data and offset, they are not cleared. and can cause corruption